umi_tools/extract.py

```diff
@@ -190,77 +219,19 @@ def main(argv=None):
-        if not options.read2_in:
-            U.error("Must supply --bc-pattern for single-end")
-        else:
-            U.error("Must supply --bc-pattern and/or --bc-pattern "
-                    "if paired-end ")
-
-    if options.pattern2:
-        if not options.read2_in:
-            U.error("must specify a paired fastq ``--read2-in``")
-
-        if not options.pattern2:
-            options.pattern2 = options.pattern
-
-    extract_cell = False
-    extract_umi = False
-
-    # If the pattern is a regex we can compile the regex(es) prior to
-    # ExtractFilterAndUpdate instantiation
-    if options.extract_method == "regex":
-        if options.pattern:
-            try:
-                options.pattern = regex.compile(options.pattern)
-            except regex.error:
-                U.error("barcode_regex '%s' is not a "
-                        "valid regex" % options.pattern)
-
-        if options.pattern2:
-            try:
-                options.pattern2 = regex.compile(options.pattern2)
-            except regex.Error:
-                U.error("barcode_regex2 '%s' is not a "
-                        "valid regex" % options.pattern2)
-
-    # check whether the regex contains a umi group(s) and cell groups(s)
-    if options.extract_method == "regex":
-        if options.pattern:
-            for group in options.pattern.groupindex:
-                if group.startswith("cell_"):
-                    extract_cell = True
-                elif group.startswith("umi_"):
-                    extract_umi = True
-        if options.pattern2:
-            for group in options.pattern2.groupindex:
-                if group.startswith("cell_"):
-                    extract_cell = True
-                elif group.startswith("umi_"):
-                    extract_umi = True
-
-    # check whether the pattern string contains umi/cell bases
-    elif options.extract_method == "string":
-        if options.pattern:
-            if "C" in options.pattern:
-                extract_cell = True
-            if "N" in options.pattern:
-                extract_umi = True
-        if options.pattern2:
-            if "C" in options.pattern2:
-                extract_cell = True
-            if "N" in options.pattern2:
-                extract_umi = True
-
-    if not extract_umi:
-        if options.extract_method == "string":
-            U.error("barcode pattern(s) do not include any umi bases "
-                    "(marked with 'Ns') %s, %s" % (
-                        options.pattern, options.pattern2))
-        elif options.extract_method == "regex":
-            U.error("barcode regex(es) do not include any umi groups "
-                    "(starting with 'umi_') %s, %s" (
-                        options.pattern, options.pattern2))
+    extract_cell, extract_umi = U.validateExtractOptions(options)
+
+    if options.either_read:
+        if extract_cell:
+            U.error("Option to extract from either read (--either-read) "
```

Comment:
Why not?

Reply:
Quite simply because I'm not sure what we would want to do with the cell barcode. Right now, we're supporting the most basic form of barcodes on either read: 

- There is a single UMI
- This could be on either read
- The correct read can be identified using regexes
- If both reads match their regex, either discard read pair, or designate one read the UMI read based on base call qualities

We could extend this to allow cell barcodes to be extracted in the same manner but I'm not sure what we're expecting? Do we expect the cell barcode and UMI are on opposite reads but could be either way around, or maybe they are on the same read? Perhaps the cell barcode is on both reads but the UMI just one, or vis versa? 

My preference would be to not support cell barcodes within the "either read" option until we know we're dealing with. Happy to put a suggestion in the error message to raise an issue on GitHub should any issue desire this functionality.

Reply:
Okay, lets see if anyone complains.